### PR TITLE
Faster bit unpacking (Part 1)

### PR DESCRIPTION
**Pull request by siddharthteotia**
Couple of improvements have been done for bit-unpacking.

- Use hand-written unpack methods for power of 2 (1, 2, 4, 8, 16, 32) number of bits used to encode the dictionaryId. The hand-written methods are faster than generic due to simplified bit math.
- Amortize the overhead of function calls.

Right now, the new code isn't yet wired into existing bit reader and writer. Couple of follow-ups will be coming soon:

- Evaluate this optimization for non power of 2 number of bits. It is fairly possible but the performance benefit of using a special hand-written function for unpacking seems to get lost as the bit math itself gets complicated with branches for non power of 2 number of bits.
- Consider using a new format where if the number of bits to encode is non power of 2, we convert it to nearest power of 2. This means if you need more than 16 bits, we use 32 bits (raw value). We get diminished returns as the overhead of unpacking itself increases at the cost of saving 10-12 bits. 
- Integrate the new changes with existing code. 

**Description of changes:**

A new version of FixedBitIntReaderWriter is written that underneath uses a new version of fast bit unpack reader PinotDataBitSetV2.

There are 3 important APIs here:

`public int readInt(int index)` 
Exists in the current code as well - Used by the scan operator to read through the forward index and dictId for each docId

`public void readInt(int startDocId, int length, int[] buffer)`
Exists in the current code as well - Used by the multi-value bit reader to get dictId for all MVs in a given cell. 

`public void readValues(int[] docIds, int docIdStartIndex, int docIdLength, int[] values, int valuesStartIndex)`
Exists at the FixedBitSingleColumnSingleValueReader interface and used by the dictionary based group by executor to get dictIds for a set of docIds (monotonically increasing but not necessarily contiguous). But the API still issued single read calls underneath. This PR introduces this API at the FixedBitIntReaderWriterV2 level so that group by executor can leverage it using the bulk read semantics.  

When this code is wired in, the scan operator will start using one of the second or third API.

Please see the [spreadsheet ](https://docs.google.com/spreadsheets/d/1mz_TQe0rXadWPtA_Xov6cXwYrSvUpQB1p1b_ZqROTDQ/edit?usp=sharing)for performance numbers.

Two kinds of tests were done:

- Compare the performance of sequential consecutive reads using single read API `getInt(index)` with faster bit unpacking code.
- Compare the performance of sequential consecutive reads using array API `readInt(int startDocId, int length, int[] buffer)` with faster bit unpacking code. 

Will be adding some units tests. The current PR has performance test.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/FixedBitIntReaderWriterV2.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import java.io.IOException;
+import org.apache.pinot.core.plan.DocIdSetPlanNode;
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+
+
+public final class FixedBitIntReaderWriterV2 implements Closeable {
+  private volatile PinotDataBitSetV2 _dataBitSet;
+  private final int _numBitsPerValue;
+
+  public FixedBitIntReaderWriterV2(PinotDataBuffer dataBuffer, int numValues, int numBitsPerValue) {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/FixedBitIntReaderWriterV2.java:32]: Consider writing a header for backward/forward compatibility.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/FixedBitIntReaderWriterV2.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    }
+  }
+
+  private boolean shouldBulkRead(int[] docIds, int startIndex, int endIndex) {
+    int numDocsToRead = endIndex - startIndex + 1;
+    int docIdRange = docIds[endIndex] - docIds[startIndex] + 1;
+    if (docIdRange > DocIdSetPlanNode.MAX_DOC_PER_CALL) {
+      return false;
+    }
+    return numDocsToRead >= ((double)docIdRange * 0.7);
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/FixedBitIntReaderWriterV2.java:85]: Reasoning for magic number?

File: pinot-perf/src/main/java/org/apache/pinot/perf/ForwardIndexBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+import org.apache.commons.math.util.MathUtils;
+import org.apache.pinot.core.io.reader.impl.v1.FixedBitSingleValueReader;
+import org.apache.pinot.core.io.util.FixedBitIntReaderWriter;
+import org.apache.pinot.core.io.util.FixedBitIntReaderWriterV2;
+import org.apache.pinot.core.io.util.FixedByteValueReaderWriter;
+import org.apache.pinot.core.io.util.PinotDataBitSet;
+import org.apache.pinot.core.io.writer.impl.v1.FixedBitSingleValueWriter;
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+
+public class ForwardIndexBenchmark {
```
Comment (mayankshriv) [inline pinot-perf/src/main/java/org/apache/pinot/perf/ForwardIndexBenchmark.java:51]: Would be good to use JMH to make benchmark more accurate.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+ */
+package org.apache.pinot.core.io.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import org.apache.pinot.core.plan.DocIdSetPlanNode;
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+
+
+public abstract class PinotDataBitSetV2 implements Closeable {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java:27]: This class needs exhaustive unit tests to ensure all cases are covered.

Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java:27]: Added several covering all possible cases. Will do another round in a follow-up

File: pinot-perf/src/main/java/org/apache/pinot/perf/ForwardIndexBenchmark.java (hunk 3 above)
Comment (siddharthteotia) [reply pinot-perf/src/main/java/org/apache/pinot/perf/ForwardIndexBenchmark.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/FixedBitIntReaderWriterV2.java (hunk 1 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/FixedBitIntReaderWriterV2.java:31]: Yes, in the follow-up when this code is wired with reader and writer (FixedBitSingleValueReader and FixedBitSingleValueWriter) and the scan operator, I will consider if the format has to be changed and bump the version and write a header. Ri

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/FixedBitIntReaderWriterV2.java (hunk 2 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/FixedBitIntReaderWriterV2.java]: I have provided details in javadoc. Let me know if that is helpful.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/FixedBitIntReaderWriterV2.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      for (int i = docIdStartIndex; i <= docIdEndIndex; i++) {
+        values[valuesStartIndex++] = _dataBitSet.readInt(docIds[i]);
+      }
+    }
+  }
+
+  private boolean shouldBulkRead(int[] docIds, int startIndex, int endIndex) {
+    int numDocsToRead = endIndex - startIndex + 1;
+    int docIdRange = docIds[endIndex] - docIds[startIndex] + 1;
+    if (docIdRange > DocIdSetPlanNode.MAX_DOC_PER_CALL) {
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/FixedBitIntReaderWriterV2.java]: I think this is coming from the previous commit. The latest version of the PE doesn't have this check.

**Comment by Jackie-Jiang** [comment]
For the benchmark, you should also compare the worst case scenario such as 3, 5, 9, 17 bits

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+        _dataBuffer.putByte(byteOffset, (byte) firstByte);
+        bitOffsetInFirstByte = numBitsLeft;
+      }
+    }
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    _dataBuffer.close();
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java:514]: Do not close the buffer (see #5400)

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/FixedBitIntReaderWriterV2.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+import com.google.common.base.Preconditions;
+import java.io.Closeable;
+import java.io.IOException;
+import org.apache.pinot.core.plan.DocIdSetPlanNode;
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+
+
+public final class FixedBitIntReaderWriterV2 implements Closeable {
+  private volatile PinotDataBitSetV2 _dataBitSet;
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/FixedBitIntReaderWriterV2.java:29]: done

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+  static final int MAX_VALUES_UNPACKED_SINGLE_ALIGNED_READ = 16; // comes from 2-bit encoding
+
+  private static final ThreadLocal<int[]> THREAD_LOCAL_DICT_IDS =
+      ThreadLocal.withInitial(() -> new int[DocIdSetPlanNode.MAX_DOC_PER_CALL]);
+
+  protected PinotDataBuffer _dataBuffer;
+  protected int _numBitsPerValue;
+
+  /**
+   * Unpack single dictId at the given docId. This is efficient
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java:38]: done

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+   * @param length length to read (number of docIds to read in the array)
+   * @param out out array to store the unpacked dictIds
+   * @param outpos starting index in the out array
+   */
+  public void readInt(int[] docIds, int docIdsStartIndex, int length, int[] out, int outpos) {
+    int startDocId = docIds[docIdsStartIndex];
+    int endDocId = docIds[docIdsStartIndex + length - 1];
+    int[] dictIds = THREAD_LOCAL_DICT_IDS.get();
+    // do a contiguous bulk read
+    readInt(startDocId, endDocId - startDocId + 1, dictIds);
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java:76]: See the outer API in FixedBitIntReaderWriterV2 that calls this. That API tries to judge sparseness before deciding to do bulk read. The decision is made on a chunk of values at a time

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+   * for each docId, a post-processing step is needed after the bulk contiguous
+   * read to correctly set the unpacked dictId into the out array throwing away
+   * the unnecessary dictIds unpacked as part of contiguous read
+   * @param docIds docIds array
+   * @param docIdsStartIndex starting index in the docIds array
+   * @param length length to read (number of docIds to read in the array)
+   * @param out out array to store the unpacked dictIds
+   * @param outpos starting index in the out array
+   */
+  public void readInt(int[] docIds, int docIdsStartIndex, int length, int[] out, int outpos) {
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java:71]: > For the benchmark, you should also compare the worst case scenario such as 3, 5, 9, 17 bits

I am still working on adding faster methods for non power of 2. Follow up will address this.

**Comment by siddharthteotia** [comment]
> For the benchmark, you should also compare the worst case scenario such as 3, 5, 9, 17 bits

I am still working on adding faster methods for non power of 2. Follow up will address this.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+  public static class Bit2Encoded extends PinotDataBitSetV2 {
+    Bit2Encoded(PinotDataBuffer dataBuffer, int numBits) {
+      _dataBuffer = dataBuffer;
+      _numBitsPerValue = numBits;
+    }
+
+    @Override
+    public int readInt(int index) {
+      long bitOffset = (long) index * _numBitsPerValue;
+      int byteOffset = (int) (bitOffset / Byte.SIZE);
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java:111]: done

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java:71]: See the outer API in FixedBitIntReaderWriterV2 that calls this. That API tries to judge sparseness before deciding to do bulk read. The decision is made on a chunk of values at a time and this bulk API is called for each chunk.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java (hunk 6 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/PinotDataBitSetV2.java]: done

**Comment by siddharthteotia** [comment]
Address the review comments. Please take another look. This is standalone code at this point so want to get this in sooner and put up follow-ups in next couple of days.

Here are the JMH results:
`Benchmark                                               	     Mode  Cnt   Score  Units
BenchmarkPinotDataBitSet.twoBitBulkWithGaps            		  	thrpt    3  28.537   ops/ms
BenchmarkPinotDataBitSet.twoBitBulkWithGapsFast        		  	thrpt    3  44.332   ops/ms
BenchmarkPinotDataBitSet.fourBitBulkWithGaps           		  	thrpt    3  28.262   ops/ms
BenchmarkPinotDataBitSet.fourBitBulkWithGapsFast       		  	thrpt    3  42.449   ops/ms
BenchmarkPinotDataBitSet.eightBitBulkWithGaps          		  	thrpt    3  26.816   ops/ms
BenchmarkPinotDataBitSet.eightBitBulkWithGapsFast      		  	thrpt    3  38.777   ops/ms
BenchmarkPinotDataBitSet.sixteenBitBulkWithGaps        		  	thrpt    3  21.095   ops/ms
BenchmarkPinotDataBitSet.sixteenBitBulkWithGapsFast    	 	  	thrpt    3  38.380   ops/ms

BenchmarkPinotDataBitSet.twoBitContiguous               	  	avgt     3  26.188   ms/op
BenchmarkPinotDataBitSet.twoBitContiguousFast           	  	avgt     3  15.692   ms/op
BenchmarkPinotDataBitSet.fourBitContiguous              	  	avgt     3  26.113   ms/op
BenchmarkPinotDataBitSet.fourBitContiguousFast          	  	avgt     3  15.178   ms/op
BenchmarkPinotDataBitSet.eightBitContiguous             	  	avgt     3  26.693   ms/op
BenchmarkPinotDataBitSet.eightBitContiguousFast          	  	avgt     3  15.726   ms/op
BenchmarkPinotDataBitSet.sixteenBitContiguous           	  	avgt     3  43.968   ms/op
BenchmarkPinotDataBitSet.sixteenBitContiguousFast       	  	avgt     3  21.390   ms/op

BenchmarkPinotDataBitSet.twoBitBulkContiguous           	  	avgt     3  32.504   ms/op
BenchmarkPinotDataBitSet.twoBitBulkContiguousFast       	  	avgt     3  14.614   ms/op
BenchmarkPinotDataBitSet.fourBitBulkContiguous          	  	avgt     3  30.794   ms/op
BenchmarkPinotDataBitSet.fourBitBulkContiguousFast      	  	avgt     3  14.583   ms/op
BenchmarkPinotDataBitSet.eightBitBulkContiguous         	  	avgt     3  16.525   ms/op
BenchmarkPinotDataBitSet.eightBitBulkContiguousFast     	  	avgt     3  10.777    ms/op
BenchmarkPinotDataBitSet.sixteenBitBulkContiguous       	  	avgt     3  54.731   ms/op
BenchmarkPinotDataBitSet.sixteenBitBulkContiguousFast   	  	avgt     3  19.312   ms/op

BenchmarkPinotDataBitSet.twoBitBulkContiguousUnaligned          avgt   	 2  32.018   ms/op
BenchmarkPinotDataBitSet.twoBitBulkContiguousUnalignedFast    	avgt     2  14.344   ms/op
BenchmarkPinotDataBitSet.eightBitBulkContiguousUnaligned        avgt     2  21.204   ms/op
BenchmarkPinotDataBitSet.eightBitBulkContiguousUnalignedFast    avgt     2  15.393   ms/op
BenchmarkPinotDataBitSet.fourBitBulkContiguousUnaligned         avgt     2  20.125   ms/op
BenchmarkPinotDataBitSet.fourBitBulkContiguousUnalignedFast     avgt     2  14.836   ms/op
BenchmarkPinotDataBitSet.sixteenBitBulkContiguousUnaligned      avgt     2  58.086   ms/op
BenchmarkPinotDataBitSet.sixteenBitBulkContiguousUnalignedFast  avgt     2  22.002   ms/op`

**Comment by siddharthteotia** [comment]
Here are the JMH results:


`

Benchmark | Score | Mode
-- | -- | --
BenchmarkPinotDataBitSet.twoBitBulkWithGaps | 28.537   ops/ms | Throughpput
BenchmarkPinotDataBitSet.twoBitBulkWithGapsFast | 44.332   ops/ms | Throughpput
BenchmarkPinotDataBitSet.fourBitBulkWithGaps | 28.262   ops/ms | Throughpput
BenchmarkPinotDataBitSet.fourBitBulkWithGapsFast | 42.449   ops/ms | Throughpput
BenchmarkPinotDataBitSet.eightBitBulkWithGaps | 26.816   ops/ms | Throughpput
BenchmarkPinotDataBitSet.eightBitBulkWithGapsFast | 38.777   ops/ms | Throughpput
BenchmarkPinotDataBitSet.sixteenBitBulkWithGaps | 21.095   ops/ms | Throughpput
BenchmarkPinotDataBitSet.sixteenBitBulkWithGapsFast | 38.380   ops/ms | Throughpput
  |   |  
BenchmarkPinotDataBitSet.twoBitContiguous | 26.188 ms/op | Avgt
BenchmarkPinotDataBitSet.twoBitContiguousFast | 15.692 ms/op | Avgt
BenchmarkPinotDataBitSet.fourBitContiguous | 26.113 ms/op | Avgt
BenchmarkPinotDataBitSet.fourBitContiguousFast | 15.178 ms/op | Avgt
BenchmarkPinotDataBitSet.eightBitContiguous | 26.693 ms/op | Avgt
BenchmarkPinotDataBitSet.eightBitContiguousFast | 15.726 ms/op | Avgt
BenchmarkPinotDataBitSet.sixteenBitContiguous | 43.968 ms/op | Avgt
BenchmarkPinotDataBitSet.sixteenBitContiguousFast | 21.390 ms/op | Avgt
  |   |  
BenchmarkPinotDataBitSet.twoBitBulkContiguous | 32.504 ms/op | Avgt
BenchmarkPinotDataBitSet.twoBitBulkContiguousFast | 14.614 ms/op | Avgt
BenchmarkPinotDataBitSet.fourBitBulkContiguous | 30.794 ms/op | Avgt
BenchmarkPinotDataBitSet.fourBitBulkContiguousFast | 14.583 ms/op | Avgt
BenchmarkPinotDataBitSet.eightBitBulkContiguous | 16.525 ms/op | Avgt
BenchmarkPinotDataBitSet.eightBitBulkContiguousFast | 10.777 ms/op | Avgt
BenchmarkPinotDataBitSet.sixteenBitBulkContiguous | 54.731 ms/op | Avgt
BenchmarkPinotDataBitSet.sixteenBitBulkContiguousFast | 19.312 ms/op | Avgt
  |   |  
BenchmarkPinotDataBitSet.twoBitBulkContiguousUnaligned | 32.018 ms/op | Avgt
BenchmarkPinotDataBitSet.twoBitBulkContiguousUnalignedFast | 14.344 ms/op | Avgt
BenchmarkPinotDataBitSet.eightBitBulkContiguousUnaligned | 21.204   ms/op | Avgt
BenchmarkPinotDataBitSet.eightBitBulkContiguousUnalignedFast | 15.393   ms/op | Avgt
BenchmarkPinotDataBitSet.fourBitBulkContiguousUnaligned | 20.125   ms/op | Avgt
BenchmarkPinotDataBitSet.fourBitBulkContiguousUnalignedFast | 14.836   ms/op | Avgt
BenchmarkPinotDataBitSet.sixteenBitBulkContiguousUnaligned | 58.086   ms/op | Avgt
BenchmarkPinotDataBitSet.sixteenBitBulkContiguousUnalignedFast | 22.002   ms/op | Avgt

`

**Comment by siddharthteotia** [comment]
Merging it. 

Follow-ups coming next:

- Wire with the reader and writer. Currently the fast methods here can be used for existing reader and writer if the index is power of 2 bit encoded. 
- Consider changing the format to Little Endian
- Consider aligning the bytes on the writer side. This will remove a few branches for 2/4 bit encoding to handle unaligned reads. It will also make it easier to add fast methods for non power of 2 bit encodings.

**Pull request merged**